src/Aspire.Hosting/Utils/PeriodicRestartAsyncEnumerable.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Aspire.Hosting.Utils;
+
+/// <summary>
+/// An <see cref="IAsyncEnumerable{T}"/> that periodically restarts an inner enumerable after a timeout.
+/// We specifically need this for the DCP resource watch enumerable, as the WatchAsync call can be
+/// unreliable when running long enough (30+ minutes). The watch stops receiving new events, but
+/// no cancellation (or other exception) is thrown. The simple fix is to periodically restart the watch
+/// with the only downside being that we'll receive a duplicate of the latest resource state every time
+/// we re-enable the watch.
+/// </summary>
+/// <typeparam name="T">The inner enumerated type</typeparam>
+internal sealed class PeriodicRestartAsyncEnumerable<T> : IAsyncEnumerable<T>
```

Comment:
Is there any way to implement this as a single loop in a function instead? A generator I mean.

Reply:
I refactored to a generator function.